### PR TITLE
Add optional DragAnchorStrategy to Draggable

### PR DESCRIPTION
**Pull request by prosac**
## Description

I am building an app that will be based around a canvas with things on it that can be connected visually. Think of node-graph editors that can be found in media creation app like blender. In my app I use a bunch of `Draggable`s on an `InteractiveViewer`. When zoomed in or out and then dragging an item leads to ugly behavior, since the "sesitive" area of the `feedback` is still based on the initial zoom level. After trying to get around that on app level and failing I realized that the actual change has to be in `Draggable` itself. This pull request introduced an injectable `DragAnchorStrategy` while not changing the default behavior, which is also extracted into a `DragAnchorStrategy`. The signature is `Offset Function(Draggable<Object> draggable, BuildContext context, Offset position);`, which gives you everything to influence the dragging appearance and behavior.

In my project I use this to adapt the scale of the `feedback` to the scale I got from the `TransformationController` of the InteractiveViewer.

```dart
Draggable(
  [...]
  dragAnchorStrategy: (widget, context, position) {
    final RenderBox renderObject = context.findRenderObject() as RenderBox;
    return renderObject.globalToLocal(position).scale(scale, scale);
  },
  [...]
)
```

This is work in progress and open for discussion.

**Right now 3 checks are failing and I do not understand the errors. Hints and help much appreciated.**

## Related Issues

Considering the size of my changes and the fact that they do not change the existing behavior, but only add new behavior, I will not open an issue for this.

[I asked a question on reddit some time ago](https://www.reddit.com/r/flutterhelp/comments/jpnlob/how_to_make_draggables_on_an_interactiveviewer/).
[I announced this one on reddit now too](https://www.reddit.com/r/FlutterDev/comments/kpgq06/making_flutter_draggable_usable_on_a_panandzoom/)

## Tests

I added the following tests:

* Drag and drop - when a dragAnchorStrategy is provided it gets called

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by prosac** [comment]
@googlebot I signed it!

**Comment by goderbauer** [comment]
@prosac Looks like the doc generation is failing.

**Comment by goderbauer** [comment]
@prosac Do you still have plans for this one?

**Comment by prosac** [comment]
yes. i found time to go on with my project that would need this change yesterday. i found a thing that could have prevented the doc generation from running through. i had reference to `_defaultDragAnchorStrategy` instead of `defaultDragAnchorStrategy` in the doc comments. just fixed it, updated flutter, merged master into my branch and ran `flutter analyze --flutter-repo`. it says "no issues found now". i pushed and now all the checks are currently running on ci. one already failed as you can see. i don't get what that check means and what i can do do fix it... out for a walk. will be here later again.

**Comment by prosac** [comment]
"Linux docs" and "docs-linux" failed again. both show the same error that is not very self explanatory ;-). @goderbauer can you give me a hint what is wrong or what I need to learn to understand the error?

**Comment by goderbauer** [comment]
From the log of those tests:

> error: unresolved doc reference [customDragAnchorStrategy], from widgets.defaultDragAnchorStrategy: (file:///tmp/flutter%20sdk/packages/flutter/lib/src/widgets/drag_target.dart:81:8)

Can you fix that?

**Comment by prosac** [comment]
i must have been blind. in the end it was not that hard to find ;-). thank you @goderbauer

**Comment by goderbauer** [comment]
Looks like a few checks are failing now. Can you take a look at those? Thanks.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:290]: suggestion:
```
A strategy that is used by this draggable to get the the anchor offset when it widget is being dragged.

The anchor offset refers to the distance between the users' finger and the [feedback] widget when this draggable is being dragged.

Defaults to [defaultDragAnchorStrategy].
```

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:293]: there is no DragAnchor.custom?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/drag_target.dart:293]: I feel with should deprecate DragAnchor and just uses the DragAnchorStrategy, and we can provide different DragAnchorStrategy(one for DragAnchor.child and one for DragAnchor.pointer)

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:290]: Your text is more straigh to the point. I will change it.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:293]: > there is no DragAnchor.custom?

Yes. That looks like a leftover from a work in progress state. Thanks for the heads up!

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:293]: Regarding deprecating DragAnchor. Yes, for the future I would also suggest that. But as a new contributor I did not want to make such a bold move and suggest that right away. First I wanted to show my alternative approach :-)

**Comment by prosac** [comment]
I still don't get the first failing check named `flutter-build`. I do not understand what I am supposed to do with the dashboard I arrive at when clicking the details link. @chunhtai can you help me?

**Comment by prosac** [comment]
ahh! it is magically green now.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:291]: sorry i had a typo
```
when it is being dragged.
```

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/drag_target.dart:291]: also we need a new line between all three sentence.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:292]:
```
draggable is dragged.
```

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:83]: This may not be a good implementation if we are going to deprecate dragAnchor in the future.

Ideally we should have two default strategy one for DragAnchor.child another for DragAnchor.pointer, developer can swap them out based on their needs.

**Review by chunhtai — COMMENTED** [review]
We should deprecates the original API as part of this PR, otherwise this API may be confusing  for the user even if it is only temporary.

you can take a look at https://github.com/flutter/flutter/wiki/Tree-hygiene#deprecation

**Comment by chunhtai** [comment]
flutter build is to make sure the master branch is in a good shape before other PR can merge, you don't need to worry about it :)

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: this was basically the bridge between my approach and the existing approach.

do you have suggestions how the interface for the user of `Draggable` would look like?

some thoughts:

```dart
Draggable(child: Whatever()) // would use DragAnchorStrategies.child like the current default.

Draggable(
  dragAnchorStrategy: DragAnchorStrategies.pointer // to choose the other build in  one
)

Draggable(
  dragAnchorStrategy: (draggable, context, position) {
    // my custom strategy
  }
)
```

what do you think. am i going in the right direction?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: The original approach uses enum which makes it less customizable. My thinking is this
```dart
Offset defaultPointerStrategy(Draggable<Object> draggable, BuildContext context, Offset position) {
  return Offset.zero;
}

Offset defaultChildStrategy(Draggable<Object> draggable, BuildContext context, Offset position) {
  final RenderBox renderObject = context.findRenderObject()! as RenderBox;
  return renderObject.globalToLocal(position);
}


Draggable(
  dragAnchorStrategy: defaultPointerStrategy // developer can choose which default they want, or provide custom function.
)
```

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: just doing

```dart
Draggable(
  dragAnchorStrategy: defaultPointerStrategy // developer can choose which default they want, or provide custom function.
)
```
wont work. `defaultPointerStrategy` is not a thing on top level or in the scope you are in when using `Draggable` somewhere.

i will implement my thoughts and push later. maybe my example was not complete enough.

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: not sure if i understand, the defaultPointerStrategy can be a static method in Draggable, so if you have access to Draggable, you should be able to use it?

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: > i will implement my thoughts and push later. maybe my example was not complete enough.
played around with it but i am too tired now to finish it.

i would rather not put it as a top level function into `drag_target.dart`. imagine you are building an app and you do 
`import 'package:flutter/material.dart';` and our new function `defaultPointerStrategy` is suddenly in top level namespace. i would not like that.

implementing it as a static method on the `Draggable`. could be a thing, but i would be more explicit and give the strategies their own namespace in the form of a class like:

```dart
class DragAnchorStrategies {
  static DragAnchorStrategy child = (Draggable<Object> draggable, BuildContext context, Offset position) {
    final RenderBox renderObject = context.findRenderObject()! as RenderBox;
    return renderObject.globalToLocal(position);
  };

  static DragAnchorStrategy pointer = (Draggable<Object> _draggable, BuildContext _context, Offset _position) {
    return Offset.zero;
  };
}

```

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: We normally do not create a class just to host static method. I think we should go with static method on draggable. We should follow the pattern we already have in flutter. For example https://api.flutter.dev/flutter/widgets/Navigator/onGenerateInitialRoutes.html

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: also i would like to express that i do not think going for a breaking interface change is a thing that we should consider now. i would suggest to introduce the new, more flexible approach and later deprecate the `dragAnchor`.

i would like to have some more opinions on this.

slightly different topic: in any case this pull request is not really ready. i realized that i did not adapt `LongPressDraggable` yet. i will check out my changes with a real android device to check out how they affect real finger style usage.

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: The reason why i think we should deprecates dragAnchor is that the behavior will be quite confusing if developers try to use both them at the same time. a custom dragAnchorStrategy will override the behavior of dragAnchor entirely. By marking it deprecated, developer can know that dragAnchorStrategy is more preferable than using dragAnchor.

Note that developer can still use deprecated API until we finally remove it after a certain deprecation period, so it is important that we do not change the behavior if developers still use the dragAnchor but not dragAnchorStrategy

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:83]: Ahh. Ok. Understood. During the weekend I will have time again and will look into how to mark things as deprecated in flutter.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [inline packages/flutter/lib/src/widgets/drag_target.dart:286]: this does not have the effect that i see it as deprecated in android studio. seems like this is not the right place. @chunhtai can you give me a hint?

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart]: @chunhtai please have another look. are my newest changes what you are suggesting?

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart]: while going on with my project that i need this for i just realized that i introduced an error with the last changes. i will investigate.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/drag_target.dart:286]: you need to add this in line 210, too
```dart
this.dragAnchor = DragAnchor.child,
```
Also the format is not correct. You can checkout other example
https://github.com/flutter/flutter/blob/fe2708c35bfccd821c62036e1daf79283c0bf2af/packages/flutter/lib/src/cupertino/text_field.dart#L266

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart]: added it to the other place and followed the style documented in the tree hygiene doc.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart]: fixed with d940c068830ac22ee0d3c4128c0ca099485ad8a8

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:286]: suggestion
```
/// Where this widget should be anchored during a drag.
///
/// This property is overridden by the [dragAnchorStrategy] if the latter is provided.
///
/// Defaults to [DragAnchor.child].
```

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:75]:
```suggestion
/// The default [DragAnchorStrategy] used when [Draggable.dragAnchor] is not set
```

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:82]:
```suggestion
/// The [DragAnchorStrategy] used when [Draggable.dragAnchor] set to
```

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:300]:
```suggestion
  /// Defaults to [childDragAnchorStrategy] if the [dragAnchor] is set to [DragAnchor.child] or [pointerDragAnchorStrategy] if the [dragAnchor] is set to [DragAnchor.pointer].
```

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/drag_target.dart:522]: nit: Why move this up?

**Review by chunhtai — COMMENTED** [review]
the analyzer is not happy.

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart]: changed

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart]: i don't know. maybe my urge to have declarations on top. moved it back.

**Review by chunhtai — APPROVED** [review]
LGTM, once the build passes

**Review by Piinks — COMMENTED** [review]
Hi @prosac welcome! 🎉 
It looks like this is just about ready to land, can you take a look at the analyzer failures?

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/drag_target.dart:212]: The analyzer is failing due to a missing 'v' in these notices. 😅 
```suggestion
      'This feature was deprecated after v2.1.0-10.0.pre.'
```

**Comment by prosac** [comment]
thank you @Piinks! i was blind :-)

File: packages/flutter/lib/src/widgets/drag_target.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/drag_target.dart:293]: Both 😝 
```suggestion
    'This feature was deprecated after v2.1.0-10.0.pre.'
```

Comment (prosac) [reply packages/flutter/lib/src/widgets/drag_target.dart:293]: aahhggrrr :-)

**Comment by prosac** [comment]
that is a good end of the day for me!

**Comment by chunhtai** [comment]
analyzer complains
```dart
info • 'dragAnchor' is deprecated and shouldn't be used. Use dragAnchorStrategy instead. This feature was deprecated after v2.1.0-10.0.pre. • dev/manual_tests/lib/drag_and_drop.dart:146:9 • deprecated_member_use
```
I believed you want to add `// ignore: deprecated_member_use` before the line

**Comment by prosac** [comment]
I hope tomorrow evening I will have time and energy, then I will tackle that.

**Comment by prosac** [comment]
My changes to the manual tests are rather naive and in a way  reveal that the this would only be _really_ complete when I would also change `LongPressDraggable`.

Should I do it

- [ ] withing this PR
- [ ] in a new PR
- [ ] not at all

?

**Comment by chunhtai** [comment]
> My changes to the manual tests are rather naive and in a way reveal that the this would only be _really_ complete when I would also change `LongPressDraggable`.
> 
> Should I do it
> 
> * [ ]  within this PR
> * [x]  in a new PR
> * [ ]  not at all
> 
> ?

I vote in a new pr

**Comment by prosac** [comment]
Cool! I will do that!

**Comment by prosac** [comment]
Had 11 hours (!) of sleep, no appointments a bank holiday in germany and finally coding energy for my private things again. I will not start to work on the new PR to bring in the new behavior also fro `LongPressDraggabe`. I don't know how much I get done today, but starting it can't be wrong.

**Comment by chunhtai** [comment]
@prosac There is a pr open to do that https://github.com/flutter/flutter/pull/79160

**Comment by chunhtai** [comment]
also as the author of that pr suggests https://github.com/flutter/flutter/pull/79160#issuecomment-812632827

Feel free to take over it if you want

**Pull request merged**
